packages/material-ui/src/Switch/Switch.js

```diff
@@ -163,67 +150,289 @@ const SwitchSwitchBase = styled(SwitchBase, {
 );
 
-const SwitchTrack = styled('span', {
+const SwitchRoot = styled('span', {
   name: 'MuiSwitch',
-  slot: 'Track',
-  overridesResolver: (props, styles) => styles.track,
-})(({ theme }) => ({
-  /* Styles applied to the track element. */
-  height: '100%',
-  width: '100%',
-  borderRadius: 14 / 2,
-  zIndex: -1,
-  transition: theme.transitions.create(['opacity', 'background-color'], {
-    duration: theme.transitions.duration.shortest,
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    const { styleProps } = props;
+
+    return {
+      ...styles.root,
+      ...(styleProps.edge && styles[`edge${capitalize(styleProps.edge)}`]),
+      ...styles[`size${capitalize(styleProps.size)}`],
+      ...styles.input,
+      ...(styleProps.color !== 'default' && styles[`color${capitalize(styleProps.color)}`]),
+    };
+  },
+})(({ styleProps }) => ({
+  display: 'inline-flex',
+  width: 34 + 12 * 2,
+  height: 14 + 12 * 2,
+  overflow: 'hidden',
+  padding: 12,
+  boxSizing: 'border-box',
+  position: 'relative',
+  flexShrink: 0,
+  zIndex: 0,
+  verticalAlign: 'middle',
+  '@media print': {
+    colorAdjust: 'exact',
+  },
+  ...(styleProps.edge === 'start' && {
+    marginLeft: -8,
+  }),
+  ...(styleProps.edge === 'end' && {
+    marginRight: -8,
+  }),
+  ...(styleProps.size === 'small' && {
+    width: 40,
+    height: 24,
+    padding: 7,
+    [`& .${switchClasses.thumb}`]: {
+      width: 16,
+      height: 16,
+    },
+    [`& .${switchClasses.switchBase}`]: {
+      padding: 4,
+      [`&.${switchClasses.checked}`]: {
+        transform: 'translateX(16px)',
+      },
+    },
   }),
-  backgroundColor:
-    theme.palette.mode === 'light' ? theme.palette.common.black : theme.palette.common.white,
-  opacity: theme.palette.mode === 'light' ? 0.38 : 0.3,
 }));
 
-const SwitchThumb = styled('span', {
+const SwitchInput = styled('input', {
   name: 'MuiSwitch',
-  slot: 'Thumb',
-  overridesResolver: (props, styles) => styles.thumb,
-})(({ theme }) => ({
-  /* Styles used to create the thumb passed to the internal `SwitchBase` component `icon` prop. */
+  slot: 'Input',
+  skipSx: true,
+})({
+  cursor: 'inherit',
+  position: 'absolute',
+  opacity: 0,
+  width: '300%',
+  height: '100%',
+  top: 0,
+  left: '-100%',
+  margin: 0,
+  padding: 0,
+  zIndex: 1,
+});
+
+/* eslint-disable react/prop-types */
+
+const SwitchThumb = styled(
+  ({ styleProps: { checked }, icon, checkedIcon, className }) => {
+    if (!checked && icon) {
+      return icon;
+    }
+
+    if (checked && checkedIcon) {
+      return checkedIcon;
+    }
+
+    return <span className={className} />;
+  },
+  {
+    name: 'MuiSwitch',
+    slot: 'Thumb',
+    overridesResolver: (props, styles) => styles.thumb,
+    shouldForwardProp: () => true,
+  },
+)(({ theme }) => ({
   boxShadow: theme.shadows[1],
   backgroundColor: 'currentColor',
   width: 20,
   height: 20,
   borderRadius: '50%',
 }));
 
+const SwitchLayout = React.forwardRef((props, ref) => {
```

Comment:
```suggestion
const SwitchLayout = React.forwardRef(function SwitchLayout(props, ref) {
```

Otherwise the component will not have a proper display name.